src/security/srv_acl.c

```diff
@@ -409,60 +379,142 @@ check_access_for_groups(struct daos_acl *acl,
 	}
 
 	if (found) {
-		if (perms_have_access(grp_perms, capas))
-			return 0;
-
-		return -DER_NO_PERM;
+		*capas = convert_perms(grp_perms);
+		return 0;
 	}
 
 	return -DER_NONEXIST;
 }
 
 static int
-check_authsys_permissions(struct daos_acl *acl,
-			  struct pool_owner *ownership,
-			  Auth__Sys *authsys, uint64_t capas)
+get_authsys_capas(struct daos_acl *acl,
+		  struct ownership *ownership,
+		  Auth__Sys *authsys,
+		  uint64_t(*convert_perms)(uint64_t),
+		  uint64_t *capas)
 {
 	int rc;
 
 	/* If this is the owner, and there's an owner entry... */
 	if (strncmp(authsys->user, ownership->user,
 		    DAOS_ACL_MAX_PRINCIPAL_LEN) == 0) {
-		rc = check_access_for_principal(acl, DAOS_ACL_OWNER, NULL,
-						capas);
+		rc = get_capas_for_principal(acl, DAOS_ACL_OWNER, NULL,
+					     convert_perms,
+					     capas);
 		if (rc != -DER_NONEXIST)
 			return rc;
 	}
 
-	rc = check_access_for_principal(acl, DAOS_ACL_USER, authsys->user,
-					capas);
+	/* didn't match the owner entry, try the user by name */
+	rc = get_capas_for_principal(acl, DAOS_ACL_USER, authsys->user,
+				     convert_perms, capas);
 	if (rc != -DER_NONEXIST)
 		return rc;
 
-	return check_access_for_groups(acl, ownership, authsys, capas);
+	return get_capas_for_groups(acl, ownership, authsys, convert_perms,
+				    capas);
+}
+
+static int
+get_auth_sys_payload(Auth__Token *token, Auth__Sys **payload)
+{
+	if (token->flavor != AUTH__FLAVOR__AUTH_SYS) {
+		D_ERROR("Credential auth flavor not supported\n");
+		return -DER_PROTO;
+	}
+
+	*payload = auth__sys__unpack(NULL, token->data.len, token->data.data);
+	if (*payload == NULL) {
+		D_ERROR("Invalid auth_sys payload\n");
+		return -DER_PROTO;
+	}
+
+	return 0;
+}
+
+static void
+filter_pool_capas_based_on_flags(uint64_t flags, uint64_t *capas)
+{
+	if (flags & DAOS_PC_RO)
+		*capas &= POOL_CAPAS_RO_MASK;
+	else if (!(*capas & POOL_CAPAS_RO_MASK) ||
+	         !(*capas & ~POOL_CAPAS_RO_MASK))
+		/*
+		 * User requested RW - if they don't have permissions for both
+		 * read and write capas, we shouldn't grant them any.
+		 */
+		*capas = 0;
+}
+
+static bool
+is_ownership_valid(struct ownership *ownership)
+{
+	if (daos_acl_principal_is_valid(ownership->user) &&
+	    daos_acl_principal_is_valid(ownership->group))
+		return true;
+
+	return false;
+}
+
+static int
+get_sec_capas_for_token(Auth__Token *token, struct ownership *ownership,
+		        struct daos_acl *acl,
```

Comment:
(style)  code indent should use tabs where possible

Reply:
Fixed